[iOS] 숙소 검색 결과 뷰모델 구현

seizze opened the pull request:
### 구현한 내용

> Issue #71 을 구현하였습니다.

- 뷰모델을 구현하고, 뷰모델이 데이터소스 역할또한 하도록 구현하였습니다.
- 뷰모델이 업데이트 될 때 노티를 보내 컬렉션뷰를 reload하도록 연결하였습니다.
- 메모리에서 해제될 때 NotificationCenter의 removeObserver를 자동으로 호출하게 될 NotificationToken을 추가하였습니다.
- NotificationCenter의 addObserver를 호출하고, NotificationToken을 생성하여 리턴하는 메서드를 extension으로 구현하였습니다.

seizze committed:
[#71] feat: 검색 결과를 갖고 있는 뷰모델 구현

seizze committed:
[#71] feat: 뷰모델이 데이터소스 역할을 하도록 구현

seizze committed:
[#71] feat: 뷰모델과 컬렉션뷰 연결

seizze committed:
[#71] feat: Notification 관련 유틸 추가

seizze committed:
[#71] feat: 뷰모델이 바뀌었을 때 컬렉션뷰 업데이트하도록 연결

ehgud0670 commented on iOS/Airbnb/Airbnb/Controllers/SearchViewController.swift:
```diff
@@ -11,12 +11,27 @@ import UIKit
 final class SearchViewController: UIViewController {
     
     @IBOutlet var filterButtons: [FilterButton]!
+    @IBOutlet weak var collectionView: UICollectionView!
+    
+    private let viewModel = SearchResultViewModel()
+    
+    private var observer: NotificationToken?
     
```
@IBOutlet 끼리 프로퍼티를 나열하고, 코드로 생성한 프로퍼티끼리 나열하는건 어떠세요?
그리고 //MARK:- 로도 표시하면 구분되고 좋을 것 같아요

```swift
//MARK:- IBOutlets
@IBOutlet var filterButtons: [FilterButton]!
@IBOutlet weak var collectionView: UICollectionView!
//MARK:- internal Properties
private let viewModel = SearchResultViewModel()
private var observer: NotificationToken?
```

seizze replied:
마크 주석은 지금까지의 프로젝트에서 쓴 곳이 없어서 일관성을 위해 계속 안쓰는 게 좋을 것 같아요 ㅠ

ehgud0670 commented on iOS/Airbnb/Airbnb/ViewModels/SearchResultViewModel.swift:
```diff
@@ -0,0 +1,40 @@
+
+final class SearchResultViewModel: NSObject {
+    typealias Key = [BNB]?
+    
```
BNB 배열을 소유한 뷰모델이니깐 BNBsViewModel라는 이름이 어떨까요?
모델데이터인 [BNB] 를 기준으로 `BNBCell`, `BNBsRequest`, `BNBsTask` 등등 다른 클래스 이름들이 모두 BNB 로 시작하니
뷰모델도 BNBs로 시작하면 무슨 모델을 소유한 뷰 모델이라서 무슨 뷰에게 데이터를 전해줄지 더 알아보기 쉬울것 같습니당

seizze replied:
`BNBsViewModel`로 변경하였습니다.

ehgud0670 commented on iOS/Airbnb/Airbnb/ViewModels/SearchResultViewModel.swift:
```diff
@@ -0,0 +1,40 @@
+import UIKit
+
+final class SearchResultViewModel: NSObject {
+    typealias Key = [BNB]?
```
이거 typealias로 모델을 고정시킨거 되게 좋은 것같아요 👍
근데 protocol ViewModelBinding은 저번 미션에도 쓰셨는데 이번에 왜 구현 안하셨는지 궁금합니다.

seizze replied:
이번엔 바인딩하는 구조가 아닌 단지 컬렉션뷰를 리로드하는 동작만 하면 되기 때문에 바인딩을 이용하지 않았습니다.

ehgud0670 replied:
그러면 associatedtype 으로 key 만 지정하면 어떨까요? 이건 단지 제안이라 안해도된다 생각하시면 넘어가겠습니다

seizze replied:
협의한 대로 다음에 필요할 때 ViewModelBinding 프로토콜을 구현해 보겠습니다.

ehgud0670 commented on iOS/Airbnb/Airbnb/ViewModels/SearchResultViewModel.swift:
```diff
@@ -0,0 +1,40 @@
+        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: BNBCell.identifier, for: indexPath) as? BNBCell else { return UICollectionViewCell() }
+        return cell
+    }
+}
+
+extension Notification.Name {
+    static let bnbsDidUpdate = Notification.Name("bnbsDidUpdate")
+}
```
Notification을 확장해서 이름들을 선언하는 것 보다
`SearchResultViewModel` 뷰모델에서 노티네임을 
```swift
enum Notification {
    static let bnbsDidUpdate = Foundation.Notification.Name("bnbsDidUpdate")
}
```
이런 식으로 가지는게 어떨까요?

옵저버 패턴은 `1:n` 관계이고 `1`인 객체가 노티네임을 가지고 있으면 
`n`들이 addObsever에 노티네임을 매개변수로 넣을 때 `1` 객체의 노티네임으로만 설정해주는게 
노티네임 상수를 고를 필요가 없어서 편할 것 같은데 어떠세요?

seizze replied:
뷰모델 안에 중첩 구조로 Notification을 선언하도록 변경하였습니다.

또한 addObserver 메서드의 기본 구현이 있는 Observable 프로토콜을 구현하고, 이를 Notification이 채택하게 해서 SearchResultViewModel.Notification.addObserver { }의 형식으로 옵저빙할 수 있도록 개선하였습니다.

ehgud0670 commented on iOS/Airbnb/Airbnb/Utils/NotificationToken.swift:
```diff
@@ -0,0 +1,23 @@
+        self.token = token
+        self.center = center
+    }
+    
+    deinit {
+        center.removeObserver(token)
+    }
```
deinit될때 저절로 observer를 제거하는거 편하고 좋네요 👍

ehgud0670 commented on iOS/Airbnb/Airbnb/Utils/NotificationToken.swift:
```diff
@@ -0,0 +1,23 @@
+
+import Foundation
+
+final class NotificationToken {
+    let token: NSObjectProtocol
+    let center: NotificationCenter
```
이거 프로퍼티들 외부에서 사용할 일 없으면 private 하게 두는게 어떨까요?

seizze replied:
private 키워드 추가하였습니다.

ehgud0670 commented on iOS/Airbnb/Airbnb/Utils/NotificationCenterExtension.swift:
```diff
@@ -0,0 +1,17 @@
+import Foundation
+
+extension NotificationCenter {
+    func addObserver(forName name: NSNotification.Name?,
+                     using block: @escaping (Any?) -> Void) -> NotificationToken {
+        let token = addObserver(forName: name, object: nil, queue: .main) { block($0.object) }
+        return NotificationToken(token: token, center: self)
+    }
+}
```
토큰 발행하고 토큰이 deinit 될때 observer를 제거하는거 진짜 신박하고 좋네요 👍
근데 원래 NotificationCenter의 addObserver의 파라미터가 `addObserver(forName: name, object: nil, queue: .main)`이니 직접 만드신 addObserver도 원형 메소드와 같이 파라미터들을 설정하고 default한 매개변수 값을 설정하는게 어떠세요?

object 를 넣어줘야 하거나, queue를 main이 아닌 다른 걸로 설정해야 하는 경우가 생길수 있으니깐여

seizze replied:
NotificationCenter Extension에서 Observable 프로토콜로 바뀌었고, 이곳에 있는 addObserver 메서드의 파라미터들에 디폴트 값들을 적용하였습니다.

ehgud0670 commented on iOS/Airbnb/Airbnb/Controllers/SearchViewController.swift:
```diff
@@ -11,12 +11,27 @@ import UIKit
     
     override func viewDidLoad() {
         super.viewDidLoad()
+        
```
`configureCollectionView`라는 메소드를 만들어서 분리하는게 어떠세요? 지금은 dataSource만 설정하지만 이후 이슈에서 delegate를 설정하는것 등등 collectionView 관련 설정 코드들을 해당 함수에서 해놓으면 보기 좋을 것 같습니당

ehgud0670 reviewed (commented):
고생하셨습니다. 😀

seizze committed:
[#71] refactor: Notification 방식 개선

- Observable 프로토콜 구현
- Observable을 채택한 Notification을 뷰모델에 중첩 구조로 선언

seizze committed:
[#71] refactor: 뷰모델 네이밍 변경

seizze committed:
[#71] refactor: Access control 키워드 추가

- NotificationToken의 프로퍼티들에 private 키워드 명시

seizze committed:
[#71] refactor: addObserver에 디폴트 파라미터 추가

seizze committed:
[#71] refactor: configureCollectionView 메서드로 분리

seizze committed:
[#71] fix: 잘못된 noti를 전달하고 있던 부분 수정

seizze committed:
[#71] chore: Resolve conflicts

ehgud0670 reviewed (approved):
수고하셨습니다! 🎉
